### PR TITLE
feat(breaking): the Next.js client now uses SWR

### PR DESCRIPTION
**Pull request by Pagebakers**
#### Checklist

- [x] run `make test`
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Pagebakers** [comment]
Fixes ENG-696

**Comment by Pagebakers** [comment]
@StarpTech 

A few design considerations here.

There is the new `createWunderGraphNext` function that initializes the SWR hooks api for Nextjs.

```ts
import { createWunderGraphNext } from '@wundergraph/nextjs'
const { client, withWunderGraph, useQuery, useMutation, useSubscription, useUser, useAuth, useFileUpload } = createWunderGraphNext({
    client: new Client({baseURL: 'http://localhost:9991'}),
    ssr: true,
});
```

It will also be autogenerated, including the client. So you can do this.

```ts
import { withWunderGraph } from './generated/nextjs'

const App = ({ Component, pageProps }) => {
  return <Component {...pageProps />
}

export default withWunderGraph(App)
```

SSR can also be disabled per query.
```ts
export default function Home() {
  const weather = useQuery({ operationName: 'Weather', ssr: false })
}
```

1. Should SSR be enabled by default?
2. Should it be configurable as a parameter of `withWundergraph` or would setting it in  `createWunderGraphNext` good enough.
3. Any other api considerations?

**Comment by jensneuse** [comment]
> @StarpTech
> 
> A few design considerations here.
> 
> There is the new `createWunderGraphNext` function that initializes the SWR hooks api for Nextjs.
> 
> ```ts
> import { createWunderGraphNext } from '@wundergraph/nextjs'
> const { withWunderGraph, useQuery, useMutation, useSubscription, useUser, useAuth, useFileUpload } = createWunderGraphNext({
>     client: new Client({baseURL: 'http://localhost:9991'}),
>     ssr: true,
> });
> ```
> 
> It will also be autogenerated, including the client. So you can do this.
> 
> ```ts
> import { withWunderGraph } from './generated/nextjs'
> 
> const App = ({ Component, pageProps }) => {
>   return <Component {...pageProps />
> }
> 
> export default withWunderGraph(App)
> ```
> 
> SSR can also be disabled per query.
> 
> ```ts
> export default function Home() {
>   const weather = useQuery({ operationName: 'Weather', ssr: false })
> }
> ```
> 
> 1. Should SSR be enabled by default?
> 2. Should it be configurable as a parameter of `withWundergraph` or would setting it in  `createWunderGraphNext` good enough.
> 3. Any other api considerations?

The API looks good to me.
1. Enabling SSR by default is ok if I can easily opt-out globally (possible?)
2. maybe make it an optional arg in `withWunderGraph` to disable it?
3. LGTM

File: packages/nextjs/src/ssr-middleware.test.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import { SWRConfig } from 'swr';
+import { WunderGraphProvider } from './context';
+import { SSRMiddleWare } from './ssr-middleware';
+
+const createClient = (overrides?: Partial<ClientConfig>) => {
+	return new Client({
+		sdkVersion: '1.0.0',
+		baseURL: 'https://api.com',
+		applicationHash: '123',
+		applicationPath: 'app',
```
Comment (jensneuse) [inline packages/nextjs/src/ssr-middleware.test.tsx]: is this not deprecated?

**Review by jensneuse — APPROVED** [review]

**Review by jensneuse — CHANGES_REQUESTED** [review]
should this be updated to the new /app/main structure already?

File: packages/nextjs/src/ssr-middleware.test.tsx (hunk 1 above)
Comment (Pagebakers) [reply packages/nextjs/src/ssr-middleware.test.tsx]: applicationPath is deprecated, but not part of this PR.

File: packages/nextjs/src/ssr-middleware.test.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+): ReturnType<typeof _renderWithConfig> => {
+	return _renderWithConfig(element, { ...config }, value);
+};
+
+const nockQuery = (operationName = 'Weather', wgParams = {}) => {
+	return nock('https://api.com')
+		.matchHeader('accept', 'application/json')
+		.matchHeader('content-type', 'application/json')
+		.matchHeader('WG-SDK-Version', '1.0.0')
+		.get('/app/operations/' + operationName)
```
[the remark this answers is not in the document]
Comment (Pagebakers) [reply packages/nextjs/src/ssr-middleware.test.tsx]: Idem

**Comment by fiam** [comment]
> should this be updated to the new /app/main structure already?

I will take care of that in the other PR. We can merge this one first.

**Review by fiam — APPROVED** [review]
LGTM

**Review by jensneuse — APPROVED** [review]

File: docs/migrations/nextjs-0.3.0-0.4.0.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+| Version range   | Migration complexity | Info                |
+| --------------- | -------------------- | ------------------- |
+| swr-0.3.0-0.4.0 | medium               | Updated to SWR 2.0. |
+
+Next.js 0.4.0 has been fully rewritten and now uses SWR to manage client server state.
+
+## Dependencies
+
+SWR is a new peer dependency of `@wundergraph/nextjs`, add it to your project.
```
Comment (StarpTech) [inline docs/migrations/nextjs-0.3.0-0.4.0.md:11]: peerDep?

File: docs/migrations/nextjs-0.3.0-0.4.0.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+});
+
+trigger({
+  name: 'Never gonna give you up',
+});
+```
+
+### useLiveQuery
+
+useLiveQuery no longer exists, you can pass `liveQuery: true` to useQuery to turn it into a live query (if it is supported)
```
Comment (StarpTech) [inline docs/migrations/nextjs-0.3.0-0.4.0.md:79]: In what case, it is not supported?

File: docs/migrations/nextjs-0.3.0-0.4.0.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+### useSubscription
+
+```ts
+const { response, refetch } = useQuery.TopProducts({
+  input: {
+    limit: 100,
+  },
+});
+
+// response.data
```
Comment (StarpTech) [inline docs/migrations/nextjs-0.3.0-0.4.0.md:100]: Remove for consistency?

File: packages/nextjs/src/types.ts [hunk 6]
```diff
@@ -1,169 +1,49 @@
...
-	files: FileList;
-	abortSignal?: AbortSignal;
+export interface WithWunderGraphOptions {
+	/**
+	 * The WunderGraph client.
+	 */
+	client: Client;
+	/**
+	 * The cache key used for user data.
+	 * @default wg_user
```
Comment (StarpTech) [inline packages/nextjs/src/types.ts:28]: What is wg_user?

File: packages/nextjs/src/with-wundergraph.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+						Cookie: cookieHeader,
+					});
+				}
+
+				let ssrUser: User | null = null;
+
+				if (fetchUserSSR !== false) {
+					try {
+						ssrUser = await client.fetchUser();
+					} catch (e) {}
```
Comment (StarpTech) [inline packages/nextjs/src/with-wundergraph.tsx:75]: Sure, that we don't want to log something e.g only in development?

File: packages/nextjs/src/ssr-middleware.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+					throw result.error;
+				}
+				return result.data;
+			};
+		}
+
+		if (ssrCache && !ssrCache[_key] && ssrFetcher && !shouldAuthenticate) {
+			// We swallow errors here, since there is no official way to pass errors back to the SWR fallback.
+			// Unless we use our own cache instance, but that might break compatibility with other apps.
+			ssrCache[_key] = ssrFetcher(key).catch(() => {});
```
Comment (StarpTech) [inline packages/nextjs/src/ssr-middleware.ts:56]: Same here. Log in development only?

**Review by StarpTech — CHANGES_REQUESTED** [review]
Overall very great only a few questions and possible improvements.

File: docs/migrations/nextjs-0.3.0-0.4.0.md (hunk 3 above)
Comment (Pagebakers) [reply docs/migrations/nextjs-0.3.0-0.4.0.md:11]: Moved it to peerDeps to make sure projects use 1 and the same version and prevent any unwanted issues with context. 

SWR 2 has these exports which a project might use.

```
import useSWRInfinite from 'swr/infinite'
import useSWRMutation from 'swr/mutation'
import useSWRImmutable from 'swr/immutable'
```

File: docs/migrations/nextjs-0.3.0-0.4.0.md (hunk 4 above)
Comment (Pagebakers) [reply docs/migrations/nextjs-0.3.0-0.4.0.md:79]: Not sure what you mean? `useLiveQuery` hook is no longer exported and you need to use `useQuery` with `liveQuery: true` instead.

File: docs/migrations/nextjs-0.3.0-0.4.0.md (hunk 3 above)
Comment (StarpTech) [reply docs/migrations/nextjs-0.3.0-0.4.0.md:11]: ok

File: docs/migrations/nextjs-0.3.0-0.4.0.md (hunk 4 above)
Comment (StarpTech) [reply docs/migrations/nextjs-0.3.0-0.4.0.md:79]: > (if it is supported)

What do you mean by that?

File: packages/nextjs/src/ssr-middleware.ts (hunk 8 above)
Comment (Pagebakers) [reply packages/nextjs/src/ssr-middleware.ts]: This has been updated now.

File: packages/nextjs/src/types.ts (hunk 6 above)
Comment (StarpTech) [reply packages/nextjs/src/types.ts:28]: ok

File: docs/migrations/nextjs-0.3.0-0.4.0.md (hunk 5 above)
Comment (StarpTech) [reply docs/migrations/nextjs-0.3.0-0.4.0.md:100]: ok

File: packages/nextjs/src/ssr-middleware.ts (hunk 8 above)
Comment (StarpTech) [reply packages/nextjs/src/ssr-middleware.ts]: ok

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**